### ISSUE TITLE
Sources to only stage what is required for building, omitting large parts of VCS history

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#376]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/376)
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 23, 2018, 13:06

In order to open the doors to caching build trees, and to staging in a build farm scenario over CAS, it is important to reduce the amount of data which is staged at build time.

This issue blocks #311 **Ammend**: as per comments, this issue no longer blocks #311)

Here is a list of things which must be completed:

### Source API docs

The `Source.stage()` API documentation must be sneakily enhanced to state that only the minimum amount of data must be staged.

This can mean reducing the amount of history in `.git/`, or excluding completely the data in `.bzr` for `bzr` plugins.

### git plugin

For `git` plugins, we must allow build scripts to correctly run `git describe`, because this is being used increasingly instead of statically committed versioning data in build scripts.

The `Source.stage()` implementation must be fixed to only include:

* The latest tagged object
* All of the commits *since* the latest tagged object, up until the *ref* commit

If there are no tags, then only the `ref` commit can be included, such that `git describe --always` still works.

For `Source.init_workspace()`, the full history must be included, and upstream repo origin adjusted to point to the real upstream URL.

The `git` source plugin documentation should clearly state that this is what is going to happen for builds, and that builds using the `git` source plugin cannot expect to have additional support from the `.git/` subdirectory than just the strict minimum needed to satisfy `git describe`

### bzr plugin

For `bzr` plugin, just omit the `.bzr` directory entirely at `Source.stage()` time.

For `Source.init_workspace()`, the full history must be included, and upstream repo origin adjusted to point to the real upstream URL.

The `bzr` source plugin documentation should clearly state that this is what is going to happen for builds, and that builds using the `bzr` source plugin do not have access to the `.bzr/` subdirectory and revision history.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 23, 2018, 13:12

> If there are no tags, the `.git` directory can be completely omitted.

Shouldn't we still include the `ref` commit in that case? `git describe --always` should still work.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 23, 2018, 13:26

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 23, 2018, 13:28

> Shouldn't we still include the `ref` commit in that case? `git describe --always` should still work.

Agreed, amended.

I was working under the assumption that if there are no tags, then it means the author of the project is clearly not using `git describe`, I could be wrong about this, though.

Further, we're going to great length already to get a nicely tailored history that works for `git describe`, it's not more effort to get this working for absent tags as well.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 25, 2018, 10:27

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) 
Please can you expand on what you mean by "Ref commit"?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Apr 25, 2018, 10:40

> [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb)
> Please can you expand on what you mean by "Ref commit"?

The commit specified in the .bst file with the `ref:` key.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 30, 2018, 15:06

I guess this will have some relation with #261? Only leaving a comment here in case

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 30, 2018, 15:50

[[Gitlab user @jjardon]](https://gitlab.com/jjardon) no not related.

#261 is about saving on download time and local source cache size, this issue is strictly about what we stage in build directories, which is unacceptably large if we are going to start caching the build trees in built artifacts.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on Apr 30, 2018, 16:08

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) rigth, sorry for the noise

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 1, 2018, 12:11

mentioned in commit 1cd722c0d5f0b68a44ec24bb9a0cd428aeabab37

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 4, 2018, 15:08

mentioned in commit a0a658e109f68145f9c59bd69be2f43724e1fc3d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 9, 2018, 12:41

This is a list i've written of the series of changes that need to be made.
If i've missed anything or if anything is wrong, i'd appreciate feedback :)

**Git Plugin**

* [x]  source.stage() Made to only include the last tagged object and all the commits since then until the `ref` commit (tracking sha) 
* [x]  If no tags exist, only return the ref commit (tracking sha) 
* [x]  For source.init_workspace(): clone full history for that branch and point it at upstream (not local copy) 
* [ ]  Modify git source plugin docs 

**Bzr Plugin**

* [x]  Omit .bzr entirely in source.stage() 
* [x]  For source.init_workspace(): clone full history for that branch and point it at upstream (not local copy) 
* [ ]  Modify bzr plugin docs 


**Remote Artifacts Issue**

* [ ]  Create seperate “blob” in the remote artifact so that the cached build tree can be downloaded or ignored at whim 
* [x]  Add a flag to bst build and bst pull to prevent downloading of cached build tree 
* [x]  Change existing `no-cache` flag to be an option, allowing for separate functionality if a remote cache is used

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 10, 2018, 08:07

> Git Plugin
> * Add `describe` function to sources/git.py 

No, not sure what this is for.

The point is this: *"When the git plugin stages sources, it must stage exactly the minimal amount of history in the staged clone such that `git describe` will work in that checkout"*

The aim is: Don't stage all 10GB of history just for the relatively small amount of source code needed to build.

The problem is: More and more, softwares are relying on `git describe` to inform the build scripts what the version is, this is practical for them because:

* They dont need to encode the version in many places, they just rely on the git tag
* When a user does not use an exact release, but a git checkout, their `--version` output reveals a very precise version of what they are using.


So to summarize: We dont want to stage the whole history into the checkout area, but build scripts are going to themselves call `git describe` in the build environment, we cannot help that. So, we have to stage the minimum amount of history such that their `git describe` commands still succeed to detect the correct version.

I was informed by git folks that the last tagged object up until the branch tip should be enough history for `git describe` to function correctly.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 09:03

mentioned in commit 8931c0b5d699ce4fbec462eb8bcbf7698f17c02b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 11:26

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) So you do not want a function for calling git describe?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 10, 2018, 12:01

> [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) So you do not want a function for calling git describe?

No, what would we do with one ?

Its the payload at build time, in the sandbox, which needs to successfully call `git describe`

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 13:03

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) My thoughts were that, if git describe was going to be called anyway, it might be worth us having the functionality pass through buildstream, especially since we're modifying things that could potentially stop it from working.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 10, 2018, 14:54

> especially since we're modifying things that could potentially stop it from working.

We're specifically taking care of staging exactly the amount of history which `git` will require to report the correct `git describe` output.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 15:13

mentioned in commit a55cf2fe411efba10b8f20693bd5932ea68fe313

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 23, 2018, 11:40

assigned to [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 24, 2018, 11:39

Additional thing to add to the TODO list:
workspace reset functionality to be modified to only do a fetch if required, preventing both the cache and the fetch taking place unless explicity requested

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on May 24, 2018, 12:49

marked this issue as related to #311

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 25, 2018, 11:30

mentioned in merge request !474

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 5, 2018, 16:14

mentioned in merge request !482

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Jun 22, 2018, 12:19

Orthogonal thought.  We are now jumping through hoops to have build systems that rely on 'git describe' to work properly.  How about we by default _not_ stage any of the VCS information, and we make the staging of git information to support describe an explicit request.  In all other cases, where build systems don't rely on git describe, the amount staged is truly minimal.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 3, 2018, 16:06

[[Gitlab user @sstriker]](https://gitlab.com/sstriker) 
A new issue has been created for removing the history entirely or keeping the whole history.
https://gitlab.com/BuildStream/buildstream/issues/455

Additional information will be added as decisions are made regarding the exact implementation details

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 5, 2018, 08:49

This topic will be discussed further at GUADEC and hopefully a solution will bee agreed upon

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jul 9, 2018, 10:23

* This topic was a matter of discussion at the GUADEC Monday morning BoF session. [[Gitlab user @richardmaw]](https://gitlab.com/richardmaw)\-codethink made a description ot *the state of art*. 
* BoF minutes at https://etherpad.gnome.org/p/buildstream-bof

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Jul 9, 2018, 10:42

Could you please include the relevant bof notes directly in the issue?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 9, 2018, 11:53

Notes taken whilst being part of the discussion so excuse the rough nature of them:

## VCS reduction 

* Richard makes a description of the state of the art.
* One motivation is storing failed builds in the cache, and to re-use previous build trees.
* The motivation for storing build trees in the cache is to support incremental builds
* Patches for caching the build tree and caching failed builds are currently blocked on this.
* Issue is not able to re-use VCS history. 
* Experiments how to store minimal git history for git describe: not pretty
* Another option could be to have a SourceTransform plugin to store the `git describe` output somewhere
* Upstream bug report: for git history reduction: https://gitlab.com/BuildStream/buildstream/issues/376
* Bug report for VCS history removal as a default - https://gitlab.com/BuildStream/buildstream/issues/455
* Second issue there was raised in the hope that it would be an easy fix to un-block the caching build trees and failed build features, but this is not seen as correct. In fact it was unclear to some people why the second issue was raised in the first place. To be clear: git history reduction is still considered the blocker to progressing with the caching features mentioned above. 
* Concerns about reproducibility with caching
* Will this change break 1.0?
* Not ok to have full history in an artifact, ever. 
* Conclusion here: we think we can move forward with git history reduction by attempting it with a shell script - problem is no-one currently able to work on the issue at the moment. To be discussed outside of the BoF.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Jul 9, 2018, 12:08

Keep in mind that this, or rather #455, is also a blocker for remote
execution.  It would be good to hear an argument why it would be incorrect
as all of the conversations that I was able to be in on the topic suggested
concensus that #455 is the correct approach.  And that only elements that
do require some VCS information staged to have their builds succeed, should
be enabled for that.

If I recall correctly the current behavior can even be considered a bug.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 18, 2018, 11:18

unassigned [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 20, 2018, 11:47

marked this issue as related to #455

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jul 20, 2018, 14:09

Again here: In order to unblock work on caching of build trees and related items at the earliest possibility, we're landing !556 which simply removes all history and dealing with the regression of `git describe` usage separately in #487.

This no longer blocks #311 or any of the related issues, let's focus on fixing that regression as a separate activity in #487.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jul 20, 2018, 14:09

closed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jul 25, 2018, 11:00

marked this issue as related to #487

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jul 25, 2018, 11:01

changed the description
